File: java/ru/eustas/twim/CodecParams.java
package ru.eustas.twim;

public class CodecParams {
  public static final int MAX_LEVEL = 7;

  private static final int MAX_F1 = 4;
  private static final int MAX_F2 = 5;
  private static final int MAX_F3 = 5;
  private static final int MAX_F4 = 5;

  public static final int MAX_PARTITION_CODE = MAX_F1 * MAX_F2 * MAX_F3 * MAX_F4;
  public static final int MAX_COLOR_CODE = 30;

  public static final int MAX_CODE = MAX_PARTITION_CODE * MAX_COLOR_CODE;

  private static final int[] COLOR_QUANT = {10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
      24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54};
  private static final int[] SCALE_STEP = {1000, 631, 399, 252, 159, 100};
  private static final int SCALE_STEP_FACTOR = 40;
  private static final int[] BASE_SCALE = {4, 9, 16, 25, 36};
  private static final int BASE_SCALE_FACTOR = 36;

  private final int[] levelScale = new int[MAX_LEVEL];
  public final int[] angleBits = new int[MAX_LEVEL];
  public final int[] colorQuant = new int[3];

  // Actually, that is image params...
  private final int width;
  private final int height;

  private int partitionCode;
  private int colorCode;

  public CodecParams(int width, int height) {
    this.width = width;
    this.height = height;
  }

  public void setColorCode(int code) {
    colorCode = code;
    int c1 = code;
    int c2 = c1;
    int c3 = c1;
    colorQuant[0] = COLOR_QUANT[c1];
    colorQuant[1] = COLOR_QUANT[c2];
    colorQuant[2] = COLOR_QUANT[c3];
  }

  public void setPartitionCode(int code) {
    partitionCode = code;
    int f1 = code % MAX_F1;
    code = code / MAX_F1;
    int f2 = code % MAX_F2;
    code = code / MAX_F2;
    int f3 = code % MAX_F3;
    code = code / MAX_F3;
    int f4 = code % MAX_F4;

    int scale = (width * width + height * height) * BASE_SCALE[f2];
    for (int i = 0; i < MAX_LEVEL; ++i) {
      levelScale[i] = scale / BASE_SCALE_FACTOR;
      scale = (scale * SCALE_STEP_FACTOR) / SCALE_STEP[f3];
    }

    int bits = 9 - f1;
    for (int i = 0; i < MAX_LEVEL; ++i) {
      angleBits[i] = Math.max(bits - i - (i * f4) / 2, 0);
    }
  }

  public int getCode() {
    return partitionCode + colorCode * MAX_PARTITION_CODE;
  }

  void setCode(int code) {
    setPartitionCode(code % MAX_PARTITION_CODE);
    code /= MAX_PARTITION_CODE;
    setColorCode(code % MAX_COLOR_CODE);
  }

  public int getLevel(int[] region) {
    final int count3 = region[region.length - 1] * 3;
    if (count3 == 0) {
      return -1;
    }

    // TODO(eustas): perhaps box area is not the best value for level calculation.
    int min_y = height + 1;
    int max_y = -1;
    int min_x = width + 1;
    int max_x = -1;
    for (int i = 0; i < count3; i += 3) {
      int y = region[i];
      int x0 = region[i + 1];
      int x1 = region[i + 2];
      min_y = Math.min(min_y, y);
      max_y = Math.max(max_y, y);
      min_x = Math.min(min_x, x0);
      max_x = Math.max(max_x, x1);
    }

    int dx = max_x - min_x;
    int dy = max_y + 1 - min_y;
    int d = dx * dx + dy * dy;
    for (int i = 0; i < MAX_LEVEL; ++i) {
      if (d >= levelScale[i]) {
        return i;
      }
    }
    return MAX_LEVEL - 1;
  }
}

File: js/CodecParams.js
import {assertFalse, forEachScan, math, newInt32Array} from './Mini.js';
import {readNumber} from "./XRangeDecoder.js";
import * as SinCos from './SinCos.js';

let /** @type{number} */ MAX_LEVEL = 7;

let /** @type{number} */ MAX_F1 = 4;
let /** @type{number} */ MAX_F2 = 5;
let /** @type{number} */ MAX_F3 = 5;
let /** @type{number} */ MAX_F4 = 5;
let /** @type{number} */ MAX_LINE_LIMIT = 63;

let /** @type{number} */ MAX_PARTITION_CODE = MAX_F1 * MAX_F2 * MAX_F3 * MAX_F4;
let /** @type{number} */ MAX_COLOR_QUANT_OPTIONS = 17;
let /** @type{number} */ MAX_COLOR_PALETTE_SIZE = 32;
let /** @type{number} */ MAX_COLOR_CODE = MAX_COLOR_QUANT_OPTIONS + MAX_COLOR_PALETTE_SIZE;

/**
 * @param{number} code
 * @return{number}
 */
let makeColorQuant = (code) => 1 + ((4 + (code & 3)) << (code >> 2));

let /** @type{number} */ SCALE_STEP_FACTOR = 40;
let /** @type{number} */ BASE_SCALE_FACTOR = 36;

/** @type{number} */
let _width;
/** @type{number} */
let _height;
/** @type{number} */
let _colorQuant;
/** @type{number} */
let _paletteSize;
/** @type{number} */
let lineLimit;
/** @const @type{!Int32Array} */
let _levelScale = newInt32Array(MAX_LEVEL);
/** @const @type{!Int32Array} */
export let angleBits = newInt32Array(MAX_LEVEL);

/** @return{number} */
export let getWidth = () => _width;
/** @return{number} */
export let getHeight = () => _height;
/** @return{number} */
export let getLineLimit = () => lineLimit;

/**
 * @param{number} w
 * @param{number} h
 * @return{void}
 */
export let initForTest = (w, h) => {
  _width = w;
  _height = h;
};

/**
 * @return{number}
 */
export let getColorQuant = () => _colorQuant;

/**
 * @return{number}
 */
export let getPaletteSize = () => _paletteSize;

/**
 * @return{number}
 */
export let getLineQuant = () => SinCos.SCALE;

/**
 * @return {number}
 */
let readSize = () => {
  let /** @type{number} */ bits = -1;
  while ((bits < 8) || (readNumber(2) > 0)) {
    bits = 8 * bits + readNumber(8) + 8;
  }
  return bits;
};

/** @return{void} */
export let read = () => {
  _width = readSize();
  _height = readSize();

  let /** @type{number} */ f1 = readNumber(MAX_F1);
  let /** @type{number} */ f2 = 2 + readNumber(MAX_F2);
  let /** @type{number} */ f3 = 10 ** (3 - readNumber(MAX_F3) / 5) | 0;
  let /** @type{number} */ f4 = readNumber(MAX_F4);
  lineLimit = readNumber(MAX_LINE_LIMIT) + 1;

  let /** @type{number} */ scale = (_width * _width + _height * _height) * f2 * f2;
  for (let /** @type{number} */ i = 0; i < MAX_LEVEL; ++i) {
    _levelScale[i] = scale / BASE_SCALE_FACTOR;
    scale = ((scale * SCALE_STEP_FACTOR) / f3) | 0;
  }

  let /** @type{number} */ bits = SinCos.MAX_ANGLE_BITS - f1;
  for (let /** @type{number} */ i = 0; i < MAX_LEVEL; ++i) {
    angleBits[i] = math.max(bits - i - (((i * f4) / 2) | 0), 0);
  }
  let /** @type{number} */ colorCode = readNumber(MAX_COLOR_CODE);
  _colorQuant = makeColorQuant(colorCode);
  _paletteSize = colorCode - MAX_COLOR_QUANT_OPTIONS + 1;
};

/**
 * @param{!Int32Array} region
 * @return{number}
 */
export let getLevel = (region) => {
  let /** @type{number} */ min_y = _height + 1;
  let /** @type{number} */ max_y = -1;
  let /** @type{number} */ min_x = _width + 1;
  let /** @type{number} */ max_x = -1;
  forEachScan(region, (y, x0, x1) => {
    min_y = min_y < y ? min_y : y;
    max_y = max_y > y ? max_y : y;
    min_x = min_x < x0 ? min_x : x0;
    max_x = max_x > x1 ? max_x : x1;
  });
  assertFalse(max_y < 0);

  let /** @type{number} */ dx = max_x - min_x;
  let /** @type{number} */ dy = max_y + 1 - min_y;
  let /** @type{number} */ d = dx * dx + dy * dy;
  let /** @type{number} */ i = 0;
  for (; i < MAX_LEVEL - 1; ++i) {
    if (d >= _levelScale[i]) {
      break;
    }
  }
  return i;
};

/**
 * @param{number} v
 * @param{number} q
 * @return{number}
 */
export let dequantizeColor = (v, q) => (255 * v / (q - 1)) | 0;

/** @const @type{number} */
export let MAX_CODE = MAX_PARTITION_CODE * MAX_COLOR_CODE;

/** @const @type{number} */
export let NODE_FILL = 0;
/** @const @type{number} */
export let NODE_HALF_PLANE = 1;
/** @const @type{number} */
export let NODE_TYPE_COUNT = 2;
